diag_manager/fms_diag_buffer.F90

```diff
@@ -0,0 +1,1566 @@
+    select type(fillval)
+    type is(real(r4_kind))
+      buff = fillval
+    class default
+      call mpp_error(FATAL, 'initialize_buffer_4d: mismatch between fill and buffer values')
+    end select
+  type is(integer(i8_kind))
+    select type(fillval)
+    type is(integer(i8_kind))
+      buff = fillval
+    class default
+      call mpp_error(FATAL, 'initialize_buffer_4d: mismatch between fill and buffer values')
+    end select
+  type is(integer(i4_kind))
+    select type(fillval)
+    type is(integer(i4_kind))
+      buff = fillval
+    class default
+      call mpp_error(FATAL, 'initialize_buffer_4d: mismatch between fill and buffer values')
+    end select
+  class default
+    call mpp_error(FATAL, 'initialize buffer_4d: buffer allocated to invalid data type, this shouldnt happen')
+  end select
+
+end subroutine initialize_buffer_4d
+
+!> @brief Initializes a buffer to a given fill value
+subroutine initialize_buffer_5d (this, fillval)
+  class(buffer5d), intent(inout) :: this !< allocated 5D buffer object
+  class(*), intent(in) :: fillval!< fill value, must be same type as the allocated buffer in this
+
+  if(.not. allocated(this%buffer)) call mpp_error(FATAL, 'initialize_buffer_5d:' // &
+      'buffer not yet allocated, allocate_buffer() must be called on this object first.')
+  ! have to check fill value and buffer types match
+  select type(buff => this%buffer)
+  type is(real(r8_kind))
+    select type(fillval)
+    type is(real(r8_kind))
+      buff = fillval
+    class default
+      call mpp_error(FATAL, 'initialize_buffer_5d: mismatch between fill and buffer values')
+    end select
+  type is(real(r4_kind))
+    select type(fillval)
+    type is(real(r4_kind))
+      buff = fillval
+    class default
+      call mpp_error(FATAL, 'initialize_buffer_5d: mismatch between fill and buffer values')
+    end select
+  type is(integer(i8_kind))
+    select type(fillval)
+    type is(integer(i8_kind))
+      buff = fillval
+    class default
+      call mpp_error(FATAL, 'initialize_buffer_5d: mismatch between fill and buffer values')
+    end select
+  type is(integer(i4_kind))
+    select type(fillval)
+    type is(integer(i4_kind))
+      buff = fillval
+    class default
+      call mpp_error(FATAL, 'initialize_buffer_5d: mismatch between fill and buffer values')
+    end select
+  class default
+    call mpp_error(FATAL, 'initialize buffer_5d: buffer allocated to invalid data type, this shouldnt happen')
+  end select
+
+end subroutine initialize_buffer_5d
+
+!> @brief Add values to 0d buffer
+!! this will just call the init routine since there's only one value
+!! input_data must match allocated type of buffer object
+subroutine add_to_buffer_0d(this, input_data)
+  class(buffer0d), intent(inout) :: this !< allocated scalar buffer object
+  class(*), intent(in)      :: input_data !< data to copy into buffer
+  if( .not. allocated(this%buffer)) call mpp_error (FATAL, 'add_to_buffer_1d: buffer not yet allocated')
```

Comment:
Please include more information in the error message such as field name.  You will be able to use it for all of the error messages in these routines.